server/graphql/sensor/src/mutations/insert.rs

```diff
@@ -0,0 +1,320 @@
+use async_graphql::*;
+use graphql_core::{
+    simple_generic_errors::{
+        DatabaseError, InternalError, RecordAlreadyExist, UniqueValueViolation,
+    },
+    standard_graphql_error::{validate_auth, StandardGraphqlError},
+    ContextExt,
+};
+use graphql_types::types::SensorNode;
+use service::{
+    auth::{Resource, ResourceAccessRequest},
+    sensor::insert::{InsertSensor, InsertSensorError as ServiceError},
+};
+
+pub fn insert_sensor(
+    ctx: &Context<'_>,
+    store_id: &str,
+    input: InsertSensorInput,
+) -> Result<InsertSensorResponse> {
+    let user = validate_auth(
+        ctx,
+        &ResourceAccessRequest {
+            resource: Resource::MutateSensor,
+            store_id: Some(store_id.to_string()),
+        },
+    )?;
+
+    let service_provider = ctx.service_provider();
+    let service_context = service_provider.context(store_id.to_string(), user.user_id)?;
+
+    match service_provider
+        .sensor_service
+        .insert_sensor(&service_context, input.into())
+    {
+        Ok(sensor) => Ok(InsertSensorResponse::Response(SensorNode::from_domain(
+            sensor,
+        ))),
+        Err(error) => Ok(InsertSensorResponse::Error(InsertSensorError {
+            error: map_error(error)?,
+        })),
+    }
+}
+
+#[derive(InputObject)]
+pub struct InsertSensorInput {
+    pub id: String,
+    pub serial: String,
+    pub name: Option<String>,
+    pub is_active: Option<bool>,
+}
+
+impl From<InsertSensorInput> for InsertSensor {
+    fn from(
+        InsertSensorInput {
+            id,
+            serial,
+            name: _,
+            is_active: _,
+        }: InsertSensorInput,
+    ) -> Self {
+        InsertSensor {
+            id,
+            serial,
+            name: None,
+            is_active: Some(true),
+        }
+    }
+}
+#[derive(SimpleObject)]
+pub struct InsertSensorError {
+    pub error: InsertSensorErrorInterface,
+}
+
+#[derive(Union)]
+pub enum InsertSensorResponse {
+    Error(InsertSensorError),
+    Response(SensorNode),
+}
+
+#[derive(Interface)]
+#[graphql(field(name = "description", type = "String"))]
+pub enum InsertSensorErrorInterface {
+    SensorAlreadyExists(RecordAlreadyExist),
+    UniqueValueViolation(UniqueValueViolation),
+    InternalError(InternalError),
+    DatabaseError(DatabaseError),
+}
+
+fn map_error(error: ServiceError) -> Result<InsertSensorErrorInterface> {
+    use StandardGraphqlError::*;
+    let formatted_error = format!("{:#?}", error);
+
+    let graphql_error = match error {
+        // Standard Graphql Errors
+        ServiceError::SensorAlreadyExists => BadUserInput(formatted_error),
```

Comment:
Just a quick brief about error handling in graphql (this should definitly be in our onboarding docs). We have standard and structured errors, standards error are the one that are return by Result and structured error are built into [schema via unions](https://blog.logrocket.com/handling-graphql-errors-like-a-champ-with-unions-and-interfaces/). 

Front end doesn't know what type of standard errors exist, and thus they are meant to be there for edge cases that are outside of apps controll and that can't be actioned on by the user or don't need to be translated, an example is internal database error, or bad user input (which is when api is used in a wrong way due to bug etc..)

Structured errors are known to front end through introspection (graphql schemas are turned to typescript with codegen), and thus we can quite easily handle the different variations of errors, translate them etc.. and if api is updated, this is straight away caught during compilation (we can even do [narrowing](https://www.typescriptlang.org/docs/handbook/2/narrowing.html) on error types ). So a structured error woudl be something that needs translating or something we can action on:

https://github.com/openmsupply/open-msupply/blob/6131f65381868ccc0ad4a125940b4b3e39491596/client/packages/system/src/Stock/Components/Repack/RepackModal.tsx#L119-L125

https://github.com/openmsupply/open-msupply/blob/6131f65381868ccc0ad4a125940b4b3e39491596/server/graphql/repack/mutations/insert.rs#L102-L106